### PR TITLE
Add LiteRT export support for PyTorch backend

### PR DESCRIPTION
**Pull request by pctablet505**
## Summary

This PR enables `model.export(filepath, format="litert")` when `KERAS_BACKEND=torch`. The TF backend path already uses `TFLiteConverter.from_keras_model()` directly; this PR adds a PyTorch backend path via `litert_torch.convert`.

**Depends on #22797** — the torch-export infrastructure and backend traceability fixes are already merged.

---

### Before (TF backend only)

```mermaid
flowchart LR
    A[Keras model] -->|TF backend| B[TFLiteConverter.from_keras_model]
    B -->|TFLite ops + SELECT_TF_OPS| C[.tflite]
    C -->|Flex delegate fallback| D[ops not in TFLite builtins]
```

### After (TF **or** PyTorch backend)

```mermaid
flowchart LR
    subgraph "TensorFlow backend (existing)"
        A[Keras model] -->|TF backend| B[TFLiteConverter.from_keras_model]
        B --> C[.tflite]
    end

    subgraph "PyTorch backend (new)"
        D[Keras model] -->|torch backend| E[torch.export.export]
        E -->|ATen ops| F[litert_torch.convert]
        F --> G[.tflite]
    end
```

The PyTorch path avoids SELECT_TF_OPS / Flex delegate entirely because `litert_torch` lowers ATen ops directly.

---

### `keras/src/export/litert.py`

```mermaid
flowchart TD
    A[export_litert] --> B{backend?}
    B -->|tensorflow| C[LiteRTExporter]
    C --> D{TFLiteConverter.from_keras_model}
    B -->|torch| E[export_litert_via_torch]
    E --> F[move model to CPU]
    F --> G[litert_torch.convert]
    G --> H[restore device]
    H --> I[return .tflite path]
```

**`export_litert_via_torch()`** does three things worth reviewing:

1. **Device handling** — moves model weights to CPU before conversion (LiteRT torch requires CPU), then restores the original device.
2. **JAX x64 isolation** — `litert_torch` imports JAX and unconditionally sets `jax_enable_x64=True`. We wrap conversion in `_preserve_jax_x64_state()` to save/restore the original setting. Without this, ~900 unrelated dtype tests fail.
3. **KWargs translation** — `_prepare_litert_kwargs()` accepts the same TFLite-style kwargs (`optimizations`, etc.) and translates them to `litert_torch.convert()` arguments. Unsupported kwargs are rejected with a clear error.

### `keras/src/models/model.py`

- Expands `format="litert"` backend check from `== "tensorflow"` to `in ("tensorflow", "torch")`
- Adds `verbose=verbose` to `export_litert()` call (was missing)
- Updates docstring to list PyTorch-backend kwargs

---

## Files changed

| File | What |
|------|------|
| `keras/src/export/litert.py` | +223 lines: `export_litert_via_torch()`, device logic, JAX x64 preservation, kwargs translation |
| `keras/src/export/litert_torch_test.py` | 22 tests: numeric parity, multi-I/O, subclass, kwargs validation, quantization |
| `keras/src/models/model.py` | Backend check expansion, `verbose` fix, docstring |
| `requirements.txt` / `requirements-torch-cuda.txt` | `litert-torch>=0.9.0` |

---

## Design decisions

**Why no custom decompositions?**
`litert_torch==0.9.0` supports `mean.dim` with `dtype` and `repeat_interleave.Tensor` natively. The MPS SDPA decomposition was removed because the export path forces CPU via `device_scope("cpu")`.

**Why batch=2 in the torch exporter?**
PyTorch specializes dimensions of size 1 to constants during `torch.export` tracing. If a dynamic batch dim has concrete value 1, export fails with "specialized it to be a constant". Using batch=2 avoids this (upstream PyTorch issue [pytorch/pytorch#176349](https://github.com/pytorch/pytorch/issues/176349)).

---

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/torch/core.py [hunk 1]
```diff
@@ -617,28 +617,50 @@ def scatter_update(inputs, indices, updates, reduction=None):
...
     updates = convert_to_tensor(updates)
 
-    python_slice = __builtins__["slice"]
+    # Convert start_indices to a Python list of ints to avoid data-dependent
+    # tensor iteration, which torch.export cannot trace. If start_indices is
+    # already a list/tuple, keep it as-is. If it's a numpy array or tensor,
+    # convert to list.
+    if not isinstance(start_indices, (list, tuple)):
+        start_indices = convert_to_tensor(start_indices, dtype="int64")
+        start_indices = start_indices.tolist()
```
Comment (hertschuh) [inline keras/src/backend/torch/core.py]: In `slice`, you were checking that the indices where ints. Here, you're converting them to tensors.

Why is it different?

File: keras/src/backend/torch/linalg.py [hunk 2]
```diff
@@ -54,7 +54,9 @@ def qr(x, mode="reduced"):
             "Expected one of {'reduced', 'complete'}. "
             f"Received: mode={mode}"
         )
-    return torch.linalg.qr(x, mode=mode)
+    # torch.linalg.qr may overwrite the input matrix.
+    # Clone to avoid side effects on the original data.
```
Comment (hertschuh) [inline keras/src/backend/torch/linalg.py]: Says who? I don't think that's true.

File: keras/src/backend/torch/linalg.py [hunk 3]
```diff
@@ -79,7 +81,10 @@ def svd(x, full_matrices=True, compute_uv=True):
 def lstsq(a, b, rcond=None):
     a = convert_to_tensor(a)
     b = convert_to_tensor(b)
-    return torch.linalg.lstsq(a, b, rcond=rcond)[0]
+    # torch.linalg.lstsq may overwrite the input matrix `a` when using
+    # the default `gelsy` driver. Clone to avoid side effects on the
+    # original data (e.g. shared-memory numpy arrays).
```
Comment (hertschuh) [inline keras/src/backend/torch/linalg.py]: Gemini thinks that's true, but I have found no evidence in the links it provided. Can you check?

File: keras/src/backend/torch/numpy.py [hunk 4]
```diff
@@ -1778,6 +1778,24 @@ def repeat(x, repeats, axis=None):
             device=get_device(),
         )
 
+    # When repeats is a scalar int and axis is specified, use
+    # unsqueeze + expand + reshape instead of repeat_interleave.
+    # This preserves static shape information during torch.export
+    # (repeat_interleave introduces unbacked symbolic dimensions)
+    # and avoids the missing aten.repeat_interleave.Tensor lowering
+    # in litert_torch.
+    if isinstance(repeats, int) and axis is not None:
```
Comment (hertschuh) [inline keras/src/backend/torch/numpy.py:1786]: Can this be extended to the case when `repeats` is a list of ints?

File: keras/src/export/torch.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        **kwargs,
+    )
+    exporter.export(filepath)
+    actual_verbose = verbose if verbose is not None else True
+    if actual_verbose:
+        io_utils.print_msg(f"Saved PyTorch ExportedProgram at '{filepath}'.")
+
+
+class TorchExporter:
+    """Exporter for the PyTorch ExportedProgram (`.pt2`) format."""
```
Comment (hertschuh) [inline keras/src/export/torch.py]: Looking at the code below, having a class doesn't buy us anything, it just adds boilerplate (the __init__).

File: keras/src/export/torch.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            lambda x: convert_spec_to_tensor(x, replace_none_number=1),
+            self.input_signature,
+        )
+        sample_inputs = tuple(sample_inputs)
+
+        device = self._get_model_device()
+        if device is not None:
+            sample_inputs = tuple(
+                t.to(device) if hasattr(t, "to") else t for t in sample_inputs
+            )
```
Comment (hertschuh) [inline keras/src/export/torch.py]: Remove, this is not needed. `convert_spec_to_tensor` already puts the tensors on the correct device.

File: keras/src/export/torch.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        return filepath
+
+    def _get_model_device(self):
+        try:
+            for v in self.model.variables:
+                if hasattr(v, "value") and hasattr(v.value, "device"):
+                    return v.value.device
+        except (AttributeError, TypeError):
+            pass
+        return None
```
Comment (hertschuh) [inline keras/src/export/torch.py]: Remove per comment above.

File: keras/src/export/torch.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        sample_inputs = tuple(sample_inputs)
+
+        device = self._get_model_device()
+        if device is not None:
+            sample_inputs = tuple(
+                t.to(device) if hasattr(t, "to") else t for t in sample_inputs
+            )
+
+        if hasattr(self.model, "eval"):
+            self.model.eval()
```
Comment (hertschuh) [inline keras/src/export/torch.py]: What is this?

File: keras/src/export/torch_test.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        return models.Model(inputs=inputs, outputs=[output1, output2])
+    raise ValueError(f"Unknown model type: {type}")
+
+
+def _to_numpy(x):
+    if hasattr(x, "detach"):
+        return x.detach().cpu().numpy()
+    if hasattr(x, "numpy"):
+        return np.array(x)
+    return np.array(x)
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: We already have `backend.convert_to_numpy`, use that.

File: keras/src/export/torch_test.py [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+def _get_torch_device():
+    from keras.src.backend.torch.core import get_device
+
+    return get_device()
+
+
+def _to_torch_tensor(np_array):
+    import torch
+
+    return torch.tensor(np_array).to(_get_torch_device())
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: Remove, use `backend.convert_to_tensor`.

File: keras/src/export/torch_test.py [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+
+def _convert_to_numpy(structure):
+    return tree.map_structure(_to_numpy, structure)
+
+
+def _get_torch_device():
+    from keras.src.backend.torch.core import get_device
+
+    return get_device()
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: Remove and just import `get_device`.

File: keras/src/export/torch_test.py [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+    def test_model_with_multiple_inputs(self):
+        model = get_model("multi_input")
+        ref_input1 = np.random.normal(size=(1, 10)).astype("float32")
+        ref_input2 = np.random.normal(size=(1, 10)).astype("float32")
+        ref_input = [ref_input1, ref_input2]
+        self._verify_export_and_inference(model, ref_input)
+
+    def test_multi_output_model_export(self):
+        import torch
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: Move to the top, it's ok to import torch in unit tests.

File: keras/src/export/torch_test.py [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+                self.assertAllClose(
+                    ref_out, loaded_out, atol=self.ATOL, rtol=self.RTOL
+                )
+        else:
+            self.assertAllClose(
+                ref_output, loaded_output, atol=self.ATOL, rtol=self.RTOL
+            )
+
+    def test_export_with_custom_input_signature(self):
+        import torch
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: Move all `import torch` to the top.

File: keras/src/export/torch_test.py [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+        inputs = layers.Input(shape=(10,))
+        x = layers.Dense(16, activation="relu")(inputs)
+        residual = layers.Dense(16)(inputs)
+        x = layers.Add()([x, residual])
+        x = layers.Activation("relu")(x)
+        outputs = layers.Dense(1)(x)
+        model = models.Model(inputs=inputs, outputs=outputs)
+
+        ref_input = np.random.normal(size=(1, 10)).astype("float32")
+        self._verify_export_and_inference(model, ref_input)
```
Comment (hertschuh) [inline keras/src/export/torch_test.py]: How is this different from the other tests? What does "residual" mean here?

File: keras/src/export/torch.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        self,
+        model,
+        input_signature=None,
+        **kwargs,
+    ):
+        if backend.backend() != "torch":
+            raise ValueError(
+                "`export_torch` is only compatible with the PyTorch backend. "
+                f"Current backend: '{backend.backend()}'."
+            )
```
Comment (hertschuh) [inline keras/src/export/torch.py]: Remove, that's already checked in `model.py`.

File: requirements-common.txt [hunk 16]
```diff
@@ -14,6 +14,7 @@ ml-dtypes
 protobuf
 tensorboard
 tensorboard-plugin-profile
+etils
```
Comment (hertschuh) [inline requirements-common.txt]: Why was `etils` added? I didn't see it used.

File: keras/src/export/litert.py [hunk 17]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
                 raise ValueError(f"Unknown converter attribute '{attr}'")
+
+
+def export_litert_via_torch(
+    model, filepath, input_signature=None, verbose=None, **kwargs
+):
+    """Export Keras model to LiteRT via PyTorch backend."""
+    actual_verbose = verbose if verbose is not None else True
+    with _preserve_jax_x64_state():
+        try:
```
Comment (hertschuh) [inline keras/src/export/litert.py:199]: Why?? This is for the Torch backend only.

File: keras/src/export/litert.py [hunk 18]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+        jax = None
+        original_x64 = None
+    else:
+        original_x64 = jax.config.jax_enable_x64
+
+    try:
+        yield
+    finally:
+        if jax is not None:
+            jax.config.update("jax_enable_x64", original_x64)
```
Comment (hertschuh) [inline keras/src/export/litert.py:313]: Why?? This is for the Torch backend only.

File: keras/src/export/litert.py [hunk 19]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+            import torch
+        except ImportError:
+            raise ImportError(
+                "To export to LiteRT with the PyTorch backend, "
+                "you must install the `litert-torch` package. "
+                "Install via: pip install litert-torch"
+            )
+
+        from keras.src.export.export_utils import convert_spec_to_tensor
+
```
Comment (hertschuh) [inline keras/src/export/litert.py]: Move imports to top.

File: keras/src/export/litert.py [hunk 20]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+        original_devices = {}
+        _move_model_to_cpu(model, original_devices, torch)
+
+        from keras.src.backend.torch.core import device_scope
+
```
Comment (hertschuh) [inline keras/src/export/litert.py]: Move imports to top.

File: keras/src/export/litert.py [hunk 21]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+                if (
+                    isinstance(obj, torch.Tensor)
+                    and not isinstance(obj, torch.nn.Parameter)
+                    and obj.device.type != "cpu"
+                ):
+                    key = ("attr", f"{layer.name}.{attr_name}")
+                    original_devices[key] = str(obj.device)
+                    setattr(layer, attr_name, obj.to("cpu"))
+    except (AttributeError, RuntimeError):
+        pass
```
Comment (hertschuh) [inline keras/src/export/litert.py]: No need for all of this. This should work (with `from keras.src.backend.torch.core import get_device`).

```python
if get_device() != "cpu":
  for v in model.variables:
    v.value.data = v.value.data.to("cpu")
```

File: keras/src/export/litert.py [hunk 22]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+                if key in original_devices:
+                    obj = getattr(layer, attr_name, None)
+                    if isinstance(obj, torch.Tensor):
+                        setattr(
+                            layer,
+                            attr_name,
+                            obj.to(original_devices[key]),
+                        )
+    except (AttributeError, RuntimeError):
+        pass
```
Comment (hertschuh) [inline keras/src/export/litert.py]:
```python
if get_device() != "cpu":
  for v in model.variables:
    v.value.data = v.value.data.to(get_device())
```

File: keras/src/export/litert.py [hunk 23]
```diff
@@ -246,3 +276,374 @@ def _apply_converter_kwargs(self, converter):
...
+            expand_shape = [-1] * x.ndim
+            expand_shape[dim + 1] = repeats
+            x = x.expand(expand_shape)
+            shape = list(self.shape)
+            shape[dim] = shape[dim] * repeats
+            return x.reshape(shape)
+
+        litert_decomp.add_pre_convert_decomp(
+            repeat_interleave_self_int, _repeat_interleave_self_int_decomp
+        )
```
Comment (hertschuh) [inline keras/src/export/litert.py]: What are all these? Is this documented somewhere?

**Review by hertschuh — CHANGES_REQUESTED** [review]
There is a lot going on here.

I think it would be easier to split this at least in two:
- the Torch export
- the LiteRT Torch export

Thanks!

File: keras/src/export/litert.py (hunk 18 above)
Comment (pctablet505) [reply keras/src/export/litert.py:313]: litert_torch imports jax internally and requires `jax_enable_x64` but after export is completed we shall restore to original value.
It is for torch backend because we are using litert_torch only with torch backend.

File: keras/src/backend/torch/linalg.py (hunk 2 above)
Comment (pctablet505) [reply keras/src/backend/torch/linalg.py]: yeah removed it

File: keras/src/export/torch.py (hunk 5 above)
Comment (pctablet505) [reply keras/src/export/torch.py]: removed it

File: keras/src/export/litert.py (hunk 23 above)
Comment (pctablet505) [reply keras/src/export/litert.py]: these were operations that were not supported earlier, when writing the export code, I think now we don't need them, and can remove it.

File: keras/src/export/litert.py [hunk 24]
```diff
@@ -246,3 +276,193 @@ def _apply_converter_kwargs(self, converter):
...
+
+
+def export_litert_via_torch(
+    model, filepath, input_signature=None, verbose=None, **kwargs
+):
+    """Export Keras model to LiteRT via PyTorch backend."""
+    from keras.src.backend.torch.core import device_scope
+    from keras.src.backend.torch.core import get_device
+    from keras.src.export.export_utils import convert_spec_to_tensor
+
```
Comment (hertschuh) [inline keras/src/export/litert.py]: I think this one can move to the top

File: keras/src/export/litert.py [hunk 25]
```diff
@@ -246,3 +276,193 @@ def _apply_converter_kwargs(self, converter):
...
+    try:
+        import litert_torch
+    except ImportError:
+        raise ImportError(
+            "To export to LiteRT with the PyTorch backend, "
+            "you must install the `litert-torch` package. "
+            "Install via: pip install litert-torch"
+        )
+
+    actual_verbose = verbose if verbose is not None else True
```
Comment (hertschuh) [inline keras/src/export/litert.py]: You already did that line 31, remove

File: keras/src/export/litert.py [hunk 26]
```diff
@@ -246,3 +276,193 @@ def _apply_converter_kwargs(self, converter):
...
+    device = get_device()
+
+    # litert_torch internally imports JAX and unconditionally enables
+    # jax_enable_x64, which breaks dtype-sensitive tests elsewhere.
+    # We preserve the original setting and restore it after conversion.
+    with _preserve_jax_x64_state():
```
Comment (hertschuh) [inline keras/src/export/litert.py:198]: Don't you have to move it before `import litert_torch` then?

File: keras/src/export/litert.py [hunk 27]
```diff
@@ -246,3 +276,193 @@ def _apply_converter_kwargs(self, converter):
...
+                if input_signature is None:
+                    input_signature = get_input_signature(model)
+
+                sample_inputs = tree.map_structure(
+                    lambda x: convert_spec_to_tensor(x, replace_none_number=1),
+                    input_signature,
+                )
+                sample_inputs = tuple(sample_inputs)
+
+                if hasattr(model, "eval"):
```
Comment (hertschuh) [inline keras/src/export/litert.py]: Why would you need this `if`? `eval` is always there.

Oh, and why do you need to call `eval()`?

File: keras/src/export/litert_torch_test.py [hunk 28]
```diff
@@ -0,0 +1,572 @@
...
+
+try:
+    import litert_torch  # noqa: F401
+    import torch  # noqa: F401
+
+    _HAS_LITERT_TORCH = True
+    if _jax is not None:
+        _jax.config.update("jax_enable_x64", _ORIG_JAX_X64)
+except (ImportError, ModuleNotFoundError):
+    _HAS_LITERT_TORCH = False
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Can you import `_preserve_jax_x64_state` from `litert.py` and use that?

File: keras/src/export/litert_torch_test.py [hunk 29]
```diff
@@ -0,0 +1,572 @@
...
+    interpreter.invoke()
+
+    outputs = [interpreter.get_tensor(d["index"]) for d in output_details]
+    return outputs if len(outputs) > 1 else outputs[0]
+
+
+def _to_numpy(x):
+    if hasattr(x, "detach"):
+        return x.detach().cpu().numpy()
+    return np.asarray(x)
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Use `backend.convert_to_numpy()`

File: keras/src/export/litert_torch_test.py [hunk 30]
```diff
@@ -0,0 +1,572 @@
...
+
+    def tearDown(self):
+        super().tearDown()
+        if self._original_jax_x64 is not None:
+            try:
+                import jax
+
+                jax.config.update("jax_enable_x64", self._original_jax_x64)
+            except ImportError:
+                pass
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: If you use `_preserve_jax_x64_state` in the tests, you shouldn't need that, right?

File: keras/src/export/litert_torch_test.py [hunk 31]
```diff
@@ -0,0 +1,572 @@
...
+        saved = sys.modules.get("litert_torch", sentinel)
+        sys.modules["litert_torch"] = None
+        try:
+            with self.assertRaises(ImportError):
+                model.export(path, format="litert")
+        finally:
+            if saved is sentinel:
+                del sys.modules["litert_torch"]
+            else:
+                sys.modules["litert_torch"] = saved
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Let's remove this, I don't think it adds value. And messing with `sys.modules` is never a good idea.

File: keras/src/export/litert_torch_test.py [hunk 32]
```diff
@@ -0,0 +1,572 @@
...
+            self.assertEqual(output.shape[0], batch_size)
+
+    def test_export_with_multiple_kwargs(self):
+        """Test export with multiple litert_torch kwargs."""
+
+        try:
+            import tensorflow as tf
+        except ImportError:
+            self.skipTest("TensorFlow required for optimizations kwarg")
+
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Remove, you can always assume TF is installed for tests

File: keras/src/export/litert_torch_test.py [hunk 33]
```diff
@@ -0,0 +1,572 @@
...
+        final_devices = [str(p.device) for _, p in model.named_parameters()]
+        self.assertEqual(original_devices, final_devices)
+
+    def test_export_with_verbose(self):
+        """Test verbose output during export."""
+
+        model = models.Sequential([layers.Dense(1, input_shape=(5,))])
+        path = os.path.join(self.get_temp_dir(), "verbose_test.tflite")
+
+        # Should print message to stdout
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py:427]: Oh, shouldn't you verify that it's working?

File: keras/src/export/litert_torch_test.py [hunk 34]
```diff
@@ -0,0 +1,572 @@
...
+        path = os.path.join(self.get_temp_dir(), "bad_extension.bin")
+
+        with self.assertRaises(ValueError):
+            model.export(path, format="litert")
+
+    def test_export_restores_jax_x64_setting(self):
+        try:
+            import jax
+        except ImportError:
+            self.skipTest("JAX is required for x64 restoration checks")
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Remove, you can always assume jax is installed for tests.

File: keras/src/export/litert_torch_test.py [hunk 35]
```diff
@@ -0,0 +1,572 @@
...
+        # Empty list should not create quant_config
+        model.export(path, format="litert", optimizations=[])
+        self.assertTrue(os.path.exists(path))
+
+    def test_export_with_normalization_layer(self):
+        """Test export with normalization layer (raw tensor attributes)."""
+
+        model = models.Sequential(
+            [
+                layers.Normalization(axis=-1, input_shape=(5,)),
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py:484]: Why? What's special about `Normalization`?

File: keras/src/export/litert_torch_test.py [hunk 36]
```diff
@@ -0,0 +1,572 @@
...
+    def setUp(self):
+        super().setUp()
+        # litert_torch requires jax_enable_x64=True to correctly generate
+        # StableHLO MLIR with i64/f64 tensors from JAX bridged lowerings.
+        # Save the prior value so tearDown can restore it exactly, rather
+        # than unconditionally forcing a particular value.
+        self._original_jax_x64 = None
+        try:
+            import jax
+
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Move import to the top, you can always assume jax is installed for tests.

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/export/litert.py (hunk 26 above)
Comment (pctablet505) [reply keras/src/export/litert.py:198]: dealt with this

File: keras/src/export/litert.py (hunk 27 above)
Comment (pctablet505) [reply keras/src/export/litert.py]: Although Keras models manage training mode via the training argument to call(), `litert_torch.convert` checks `module.training` and warns when it's True. Calling `eval()` silences that warning and ensures any PyTorch-native submodules are in inference mode.

File: keras/src/export/litert_torch_test.py [hunk 37]
```diff
@@ -0,0 +1,511 @@
...
+        ):
+            model.export(
+                path,
+                format="litert",
+                allow_custom_ops=True,
+            )
+
+    # ------------------------------------------------------------------ #
+    #  Edge cases and robustness tests
+    # ------------------------------------------------------------------ #
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: remove this comment.

File: keras/src/export/litert_torch_test.py [hunk 38]
```diff
@@ -0,0 +1,511 @@
...
+            # Verify output has correct batch dimension
+            self.assertEqual(output.shape[0], batch_size)
+
+    def test_export_with_multiple_kwargs(self):
+        """Test export with multiple litert_torch kwargs."""
+
+        try:
+            import tensorflow as tf
+        except ImportError:
+            self.skipTest("TensorFlow required for optimizations kwarg")
```
Comment (hertschuh) [inline keras/src/export/litert_torch_test.py]: Remove, `tf` is already imported at the top of this file.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Please look into the failing test. Thanks!

File: keras/src/backend/torch/core.py [hunk 39]
```diff
@@ -222,11 +222,15 @@ def convert_to_tensor(x, dtype=None, sparse=None, ragged=None):
             )
 
     # Convert to np in case of any array-like that is not list or tuple.
```
Comment (pctablet505) [inline keras/src/backend/torch/core.py:224]: fix for https://github.com/keras-team/keras/issues/22815

**Review by hertschuh — CHANGES_REQUESTED** [review]
Looks good to me!

But can you rebase (on top of your other export PR)?

**Comment by hertschuh** [comment]
Hey, somehow there are still merge conflicts.

**Review by hertschuh — APPROVED** [review]
Thanks!
